File: web/codec.js
/*
 * codec.js -- JS-декодер msgpack-схемы транспорта (P6, N2). ЗЕРКАЛО
 * `core/runtime/codec.py` -- одна и та же схема, читаемая на двух языках без
 * pickle/py-специфичных расширений (msgpack -- стандарт, декодер --
 * @msgpack/msgpack с CDN, см. index.html).
 *
 * Схема сообщения (msgpack map):
 *   { topic: string, tact: int, kind: "array"|"value", payload: ... }
 *
 * kind == "array" (каналы cube/squares):
 *   payload = { shape: [int,...], dtype: string, data: Uint8Array }
 *   dtype ∈ {"complex64","complex128","float32","float64","int32","int64","uint8"}
 *   data -- raw little-endian байты, C-порядок (row-major), для complex* --
 *   ПАРЫ (re,im) подряд на элемент (как numpy .tobytes() комплексного массива).
 *
 * kind == "value" (каналы tracks/cmd): payload -- уже примитивы (объект/массив).
 */

/** dtype-имя -> [TypedArray-конструктор, число float-компонент на элемент]. */
const DTYPE_TABLE = {
  complex64: [Float32Array, 2],
  complex128: [Float64Array, 2],
  float32: [Float32Array, 1],
  float64: [Float64Array, 1],
  int32: [Int32Array, 1],
  int64: [BigInt64Array, 1],
  uint8: [Uint8Array, 1],
};

/**
 * Декодирует один msgpack-кадр (Uint8Array/ArrayBuffer) -> {topic, tact, payload}.
 * Для kind=="array" payload -- {shape, dtype, real, imag|null} (для complex*
 * действительная/мнимая части разложены в отдельные TypedArray -- удобно для
 * Three.js/canvas без комплексной арифметики на JS-стороне).
 * Требует глобальный `MessagePack` (см. index.html, CDN @msgpack/msgpack).
 */
function decodeFrame(bytes) {
  const body = MessagePack.decode(bytes instanceof Uint8Array ? bytes : new Uint8Array(bytes));
  const { topic, tact, kind, payload } = body;
  if (kind === "array") {
    const entry = DTYPE_TABLE[payload.dtype];
    if (!entry) throw new Error(`codec.js: неизвестный dtype ${payload.dtype}`);
    const [TypedArrayCtor, nComponents] = entry;
    const raw = payload.data; // Uint8Array (little-endian, как записал numpy)
    const flat = new TypedArrayCtor(raw.buffer, raw.byteOffset, raw.byteLength / TypedArrayCtor.BYTES_PER_ELEMENT);
    let real = flat;
    let imag = null;
    if (nComponents === 2) {
      const n = flat.length / 2;
      real = new TypedArrayCtor(n);
      imag = new TypedArrayCtor(n);
      for (let i = 0; i < n; i++) {
        real[i] = flat[2 * i];
        imag[i] = flat[2 * i + 1];
      }
    }
    return { topic, tact, payload: { shape: payload.shape, dtype: payload.dtype, real, imag } };
  }
  if (kind === "value") {
    return { topic, tact, payload };
  }
  throw new Error(`codec.js: неизвестный kind ${kind}`);
}

/** Кодирует команду панель->сервер (та же схема, что `codec.encode_command` в Python). */
function encodeCommand(cmd, args) {
  return MessagePack.encode({ topic: "cmd", tact: 0, kind: "value", payload: { cmd, args } });
}

File: core/runtime/codec.py
"""codec -- ЕДИНСТВЕННОЕ место (де)кодирования msgpack-сообщений транспорта (P6, N2).

🚫 НЕ `pickle` (py-специфично, C++/JS не разберут). 🚫 НЕ `msgpack-numpy` (это тоже
py-специфичное РАСШИРЕНИЕ msgpack -- ext-типы понимает только та же библиотека на
питоне). Комплексные/вещественные массивы кодируются ЯВНОЙ схемой поверх обычных
примитивов msgpack (map/str/int/bin) -- любой язык с msgpack-библиотекой (C++,
JS/`@msgpack/msgpack`, ...) читает её без спец-плагинов.

## Схема сообщения (msgpack map, ключи -- ASCII-строки)

```
{
  "topic": str,   # имя канала: "cube" | "squares" | "tracks" | "cmd" | ...
  "tact":  int,   # номер такта (>=0); у команд панель->сервер обычно 0 (не используется)
  "kind":  str,   # "array" -- закодирован numpy-массив, "value" -- произвольные примитивы
  "payload": ...  # см. ниже, форма зависит от "kind"
}
```

`kind == "array"` (каналы `cube`, `squares`) -- `payload`:
```
{
  "shape": [int, ...],   # форма массива, C-порядок (row-major)
  "dtype": str,           # "complex64" | "complex128" | "float32" | "float64" |
                          # "int32" | "int64" | "uint8"
  "data":  bytes          # сырые байты, ВСЕГДА little-endian, ровно
                          # prod(shape) * itemsize(dtype) байт, C-порядок (`.tobytes()`)
}
```
Комплексные dtype (`complex64`=2x float32, `complex128`=2x float64) -- в msgpack
НЕТ родного complex-типа, поэтому по паре (re, im) на элемент, ЗАПИСАННЫХ ПОДРЯД
(interleaved), как в родном представлении numpy `complex64`/`complex128` -- то
есть `data` -- это ПРОСТО `arr.astype('<c8').tobytes()` (little-endian complex64),
С++/JS читает так же, как `numpy` пишет: `[re0, im0, re1, im1, ...]` float32.

`kind == "value"` (каналы `tracks`, `cmd`, ...) -- `payload` -- произвольная
УЖЕ-примитивная (str/int/float/bool/None/list/dict) вложенная структура. Кодек НЕ
умеет сериализовать numpy-массивы/dataclass'ы внутри `value` -- вызывающий код
(`scene_server.py`) обязан привести их к примитивам ДО `encode()` (см.
`SceneServer._targets_payload`), иначе `msgpack.packb` кинет `TypeError`.

## Почему единственное место

Схема -- контракт для будущего C++/GPU-продюсера (P7) и JS-панели (`web/`, тот же
`decode()` переписан на JS в `web/codec.js`, ключи/формы совпадают 1:1). Если
кодирование/декодирование появится где-то ещё -- схема неизбежно разъедется.
"""
from __future__ import annotations

from typing import Any

import msgpack
import numpy as np

# dtype-имя (строка схемы) <-> numpy dtype. Явный little-endian ('<') -- фикс
# endianness (N2): хост может быть big-endian, на проводе -- ВСЕГДА LE.
_DTYPE_TO_NAME: dict[np.dtype, str] = {
    np.dtype("<c8"): "complex64",
    np.dtype("<c16"): "complex128",
    np.dtype("<f4"): "float32",
    np.dtype("<f8"): "float64",
    np.dtype("<i4"): "int32",
    np.dtype("<i8"): "int64",
    np.dtype("<u1"): "uint8",
}
_NAME_TO_DTYPE: dict[str, np.dtype] = {name: dt for dt, name in _DTYPE_TO_NAME.items()}


def _dtype_name(dtype: np.dtype) -> str:
    le_dtype = np.dtype(dtype).newbyteorder("<")
    name = _DTYPE_TO_NAME.get(le_dtype)
    if name is None:
        raise ValueError(
            f"codec: неподдерживаемый dtype={dtype} -- разрешены {sorted(_NAME_TO_DTYPE)}"
        )
    return name


def encode(topic: str, tact: int, payload: object) -> bytes:
    """`(topic, tact, payload)` -> msgpack-байты (единственная точка кодирования, N2).

    `payload` -- либо `np.ndarray` (кодируется схемой "array"), либо уже-примитивная
    структура (кодируется схемой "value" как есть).
    """
    if isinstance(payload, np.ndarray):
        arr = np.ascontiguousarray(payload).astype(
            np.dtype(payload.dtype).newbyteorder("<"), copy=False
        )
        body: dict[str, Any] = {
            "topic": topic,
            "tact": int(tact),
            "kind": "array",
            "payload": {
                "shape": list(arr.shape),
                "dtype": _dtype_name(arr.dtype),
                "data": arr.tobytes(order="C"),
            },
        }
    else:
        body = {"topic": topic, "tact": int(tact), "kind": "value", "payload": payload}
    return msgpack.packb(body, use_bin_type=True)


def decode(raw: bytes) -> tuple[str, int, object]:
    """msgpack-байты -> `(topic, tact, payload)`, обратное к `encode` (единственная точка декода).

    Для `kind == "array"` восстанавливает `np.ndarray` (копия -- `frombuffer` над
    временным `bytes`, владение памятью не шарим наружу). Для `kind == "value"`
    возвращает `payload` как есть (примитивы msgpack -> str/int/float/bool/None/list/dict).
    """
    body = msgpack.unpackb(raw, raw=False, strict_map_key=False)
    topic = body["topic"]
    tact = body["tact"]
    kind = body["kind"]
    payload = body["payload"]
    if kind == "array":
        dtype = _NAME_TO_DTYPE.get(payload["dtype"])
        if dtype is None:
            raise ValueError(f"codec: неизвестный dtype в сообщении: {payload['dtype']!r}")
        shape = tuple(payload["shape"])
        arr = np.frombuffer(payload["data"], dtype=dtype).reshape(shape).copy()
        return topic, tact, arr
    if kind == "value":
        return topic, tact, payload
    raise ValueError(f"codec: неизвестный kind={kind!r} (ожидался 'array' или 'value')")


def encode_command(cmd: str, args: dict[str, Any]) -> bytes:
    """Команда панель->сервер: та же схема "value" под фиксированным topic="cmd".

    `payload = {"cmd": cmd, "args": args}` -- `args` обязаны быть примитивами
    (см. `commands.py::Command.to_message`).
    """
    return encode("cmd", 0, {"cmd": cmd, "args": args})


def decode_command(raw: bytes) -> tuple[str, dict[str, Any]]:
    """Обратное к `encode_command` -- `(cmd_name, args)`."""
    topic, _tact, payload = decode(raw)
    if topic != "cmd" or not isinstance(payload, dict) or "cmd" not in payload:
        raise ValueError(f"codec: сообщение не похоже на команду: topic={topic!r}, payload={payload!r}")
    return str(payload["cmd"]), dict(payload.get("args") or {})
